crystal_diffusion/models/mtp.py

```diff
@@ -0,0 +1,315 @@
+
+        block_pattern = re.compile("BEGIN_CFG\n(.*?)\nEND_CFG", re.S)
+        size_pattern = re.compile("Size\n(.*?)\n SuperCell", re.S | re.I)
+        if nbh_grade:
+            position_pattern = re.compile("nbh_grades\n(.*?)\n Energy", re.S)
+        else:
+            position_pattern = re.compile("fz\n(.*?)\n Energy", re.S)
+        energy_pattern = re.compile("Energy\n(.*?)\n (?=PlusStress|Stress)", re.S)
+        # stress_pattern = re.compile("xy\n(.*?)(?=\n|$)", re.S)  # TODO stress values
+
+        for block in block_pattern.findall(lines):
+            d = {"outputs": {}}
+            size_str = size_pattern.findall(block)[0]
+            size = int(size_str.lstrip())
+            position_str = position_pattern.findall(block)[0]
+            position = np.array(list(map(formatify, position_str.split("\n"))))
+            species = np.array(self.elements)[position[:, 1].astype(np.int64)]
+            forces = position[:, 5:8].tolist()
+
+            energy_str = energy_pattern.findall(block)[0]
+            energy = float(energy_str.lstrip())
+            # TODO add stress
+            # stress_str = stress_pattern.findall(block)[0]
+            # virial_stress = (np.array(list(map(formatify, stress_str.split()))).reshape(6,).tolist())
+            # virial_stress = [virial_stress[self.mtp_stress_order.index(n)] for n in self.vasp_stress_order]
+            d["outputs"]["energy"] = energy
+            d["num_atoms"] = size
+            d["outputs"]["position"] = position[:, 2:5].tolist()
+            d["outputs"]["forces"] = forces
+            d["outputs"]["species"] = species
+            # d["outputs"]["virial_stress"] = virial_stress
+            if nbh_grade:
+                nbh_grade_values = position[:, 8].tolist()
+                d["outputs"]["nbh_grades"] = nbh_grade_values
+            data_pool.append(d)
+
+        # originally used convert_docs from maml.utils, but it hard-coded the structure of the dataframe and does not
+        # manage nbh_grades; we use our own implementation instead
+        df = self.convert_to_dataframe(docs=data_pool)
+        return df
+
+    @staticmethod
+    def convert_to_dataframe(docs: List[Dict[str, Any]]) -> pd.DataFrame:
+        """Convert a list of docs into DataFrame usable for computing metrics and analysis.
+
+        Modified from maml.utils._data_conversion.py
+
+        Args:
+            docs: list of docs generated by mlip-3. Each doc should be structured as a dict.
+
+        Returns:
+            A DataFrame with energy, force, and nbh grades (MaxVol factor) per atom and per structure. Energy is
+            repeated for each atom and corresponds to the total energy predicted.
+        """
+        df = defaultdict(list)
+        for s_idx, d in enumerate(docs):
+            outputs = d["outputs"]
+            pos_arr = np.array(outputs["position"])
+            force_arr = np.array(outputs["forces"])
+            n_atom = force_arr.shape[0]
```

Comment:
There's a `num_atoms` field in the `docs` dictionary. A few sanity check asserts on array dimensions here would help prevent the user from shooting themselves in the foot.

Reply:
good catch. Added assert for that